### PR TITLE
[4] Remove disabled and useless buttons on empty state leaving just New

### PR DESCRIPTION
**Pull request by PhilETaylor**
### Summary of Changes

Continuing the theme of Empty State Design, remove redundant, already disabled and non-sensible buttons from toolbars when in an Empty State leaving just the New button. 

In an Empty state design you are trying to guide the user to their next step, presenting disabled/greyed out buttons and buttons that cannot yet be used goes against that. 

### Testing Instructions

Apply PR, create Empty States by deleting things, look at tool bars

### Actual result BEFORE applying this Pull Request

Before this it could be as bad as these examples (Categories example is resolved in https://github.com/joomla/joomla-cms/pull/33286)

Screenshot 2021-04-27 at 19 52 26

Screenshot 2021-04-27 at 19 53 15

### Expected result AFTER applying this Pull Request

Screenshot 2021-04-27 at 19 53 51

Screenshot 2021-04-27 at 19 53 30

### Documentation Changes Required

none

//cc @Bakual

### CODE REVIEW

**Comment by Bakual** [comment]
I would actually handle this the same as an any empty list, not differing if it's empty because of filters or empty because no items exist at all. So either hide it always when `$this->items` is empty (as it's useless then as well) or always show it.

But that is a personal opinion. I can see why you want it differently. In the end it's a UX question where I have no clue to begin with 😄

**Comment by PhilETaylor** [comment]
Joomla had a UX Project at one point ... hahahaha 

> not differing if it's empty because of filters or empty because no items exist at all

I would say the complete opposite :) :) It should differ, that's one of the reasons we changed what's in the body of the page to make it more user friendly, so that it DOES differ

**Comment by ReLater** [comment]
The rebuild button can be helpful if you don't see any categories in the backend for whatever reason to repair the database table. For me it's not useless in all cases.

The same for menus.

> to make it more user friendly

Pure matter of opinion, just BTW. A disabled button shows a possible feature that is not active at the moment. Instead of seeing it you send the people to complicated and/or endless docs pages. How many people used these for newbies confusing help screens in the past?

**Comment by PhilETaylor** [comment]
> Pure matter of opinion, just BTW. A disabled button shows a possible feature that is not active at the moment. Instead of seeing it you send the people to complicated and/or endless docs pages. How many people used these for newbies confusing help screens in the past?

You totally miss understand the concept and application of an empty state. Please go and do some research on the subject. 

> The rebuild button can be helpful if you don't see any categories in the backend for whatever reason to repair the database table. For me it's not useless in all cases.

The rebuild button is TOTALLY USELESS in the case when an empty state is shown. 

The empty state for CATEGORIES will only be shown if there **ARE NO CATEGORIES in the database table (checked by running SQL)** for the filtered extension.

No matter how many times you press the Rebuild button - It will never "repair the database" and show a category in this instance because, to get the Empty State to show THERE ARE NO CATEGORIES in the db - checked by running a SQL query. You cannot magically "find" categories by pressing rebuild, because NO CATEGORIES EXIST in the database for the filtered extension.

**Comment by brianteeman** [comment]
Factually untrue. There are rows in the db just not for that extension

**Comment by PhilETaylor** [comment]
>  There are rows in the db just not for that extension

Correct. This is what I said. This is what is meant by the phrase "for the filtered extension."

**Comment by PhilETaylor** [comment]
> The rebuild button is TOTALLY USELESS in the case when an empty state is shown.

If you were to use a Rebuild button on an Empty State it would only be messing with Categories for other extensions, and would never change the empty state. It would be more confusing, as the changes it makes would only be reflected if you went to those other extensions categories pages (which would not be empty states) to see the changes. 

No matter how many times you press Rebuild button on an empty state category view, you will NEVER see a change and the empty state will NEVER change magically to show categories for that filtered extension.

**Comment by Quy** [comment]
With no categories or 1 unpublished category, do you know why there is no New/Add button? It is happening with News Feeds too. Is it dependent on another PR?

File: administrator/components/com_banners/src/View/Banners/HtmlView.php [hunk 1]
```diff
@@ -134,19 +142,19 @@ public function display($tpl = null): void
...
-		$user  = Factory::getUser();
+		$user  = Factory::getApplication()->getIdentity();
 
 		// Get the toolbar object instance
 		$toolbar = Toolbar::getInstance('toolbar');
 
 		ToolbarHelper::title(Text::_('COM_BANNERS_MANAGER_BANNERS'), 'bookmark banners');
 
-		if (count($user->getAuthorisedCategories('com_banners', 'core.create')) > 0)
+		if (\count($user->getAuthorisedCategories('com_banners', 'core.create')) > 0)
```
Comment (PhilETaylor) [inline administrator/components/com_banners/src/View/Banners/HtmlView.php:152]: @Quy I think its this...

**Comment by brianteeman** [comment]
Thats a problem because you do NOT need to have an existing category before you can create a banner

File: administrator/components/com_banners/src/View/Banners/HtmlView.php (hunk 1 above)
Comment (PhilETaylor) [reply administrator/components/com_banners/src/View/Banners/HtmlView.php:152]: Actually its not this.

**Comment by PhilETaylor** [comment]
This is a bug in the original com_banners empty state PR, and not in this PR.

**Comment by Quy** [comment]
A category is required to create a new banner.

**Comment by PhilETaylor** [comment]
> A category is required to create a new banner.

Correct. This is totally unrelated to my work on empty states. 

The code clearly states a cateogory needs to be there before the Add button in the toolbar will show.

```php
	if (count($user->getAuthorisedCategories('com_banners', 'core.create')) > 0)
		{
			$toolbar->addNew('banner.add');
		}

```

Therefore this was a preexisting issue with Joomla.

**Comment by brianteeman** [comment]
@quy a category does NOT need to be created before you create a banner. It is created automatically. Just as it is with all similar components that use categories

**Comment by PhilETaylor** [comment]
If I revert my code to the last Beta Issued of Joomla 4 and delete my banner category - I get no new button. Thus confirming nothing to do with Empty State :) - phew :) 

Screenshot 2021-04-28 at 22 56 23

**Comment by PhilETaylor** [comment]
> @Quy a category does NOT need to be created before you create a banner. It is created automatically. Just as it is with all similar components that use categories

No. it. is. not. Please test and you will see. The NEW button doesn't even show if there are no categories.

**Comment by joomdonation** [comment]
Change this line of code https://github.com/joomla/joomla-cms/blob/4.0-dev/administrator/components/com_banners/src/View/Banners/HtmlView.php#L144 to 

```php
if ($canDo->get('core.create') || count($user->getAuthorisedCategories('com_banners', 'core.create')) > 0)
```
will fix the problem.

**Comment by PhilETaylor** [comment]
You CAN create a NEW category when saving a NEW banner - but you cannot get to the NEW Banner Edit page if there are ZERO categories.

**Comment by PhilETaylor** [comment]
That line 144 has been there for 10 years , this "bug" has been dormant for 10 years. LMAO

Screenshot 2021-04-28 at 23 00 06

**Comment by PhilETaylor** [comment]
That bug is now fixed in https://github.com/joomla/joomla-cms/pull/33399

**Comment by Quy** [comment]
If you're up for it, news feeds has the same issue. Thanks.

**Comment by ghost** [comment]
@PhilETaylor 

1. After deleting all articles the `Empty Trash`-button is shown . Cause it was the last filter set and creating now a new article give them status "trashed":

> image

2. These 9 views are to test?
> image

**Comment by PhilETaylor** [comment]
1 thanks I'll fix that today
2 I forget how many empty states have been merged already - but all of them probably need the empty trash button fix and all need testing

**Comment by PhilETaylor** [comment]
> If you're up for it, news feeds has the same issue. Thanks.

Done https://github.com/joomla/joomla-cms/pull/33405

**Comment by PhilETaylor** [comment]
> After deleting all articles the Empty Trash-button is shown . Cause it was the last filter set and creating now a new article give them status "trashed":

This is now resolved and only effected Articles, Featured, Messages and Notes view models. Thanks

**Comment by Quy** [comment]
I have tested this item :white_check_mark: successfully on 35cde2eadf090d899ede3fd9de937fc84861ca64<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/33380">issues.joomla.org/tracker/joomla-cms/33380</a>.</sub>

**Comment by brianteeman** [comment]
@infograf768 always objected in the past to removing buttons and any attempt at progressive disclosure of toolbar functions

**Comment by PhilETaylor** [comment]
Joomla has a long history of bad UX decisions. It's time to move on. 

Empty state design is now a part of joomla 4. It should minimise everything on the screen to guide a user to a CTA, which is nearly always to create something with a new button, or to get help/documentation on what the future is￼￼￼

It is not really progressive disclosure, because as soon as you have one of an item, the empty state is removed revealing the full interface. If we really did have progressive disclosure of toolbar patterns, then there would never be any grayed out buttons, they would appear when they could be used in context￼

image

**Comment by ghost** [comment]
I have tested this item :white_check_mark: successfully on 35cde2eadf090d899ede3fd9de937fc84861ca64<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/33380">issues.joomla.org/tracker/joomla-cms/33380</a>.</sub>

**Comment by alikon** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/33380">issues.joomla.org/tracker/joomla-cms/33380</a>.</sub>

**Comment by richard67** [comment]
@PhilETaylor This PR here has now a merge conflict in file `administrator/components/com_tags/src/View/Tags/HtmlView.php` after your other PR #33286 has been merged. I could try to solve that, but I think it's better if you do that since you are author of both PR's, this here and the other one.

**Comment by PhilETaylor** [comment]
No worries - I was expecting conflicts as each PR was designed to be standalone and contain everything they needed to function, for ease of testing and acceptance. I'll fix this one up now.

File: administrator/components/com_tags/src/View/Tags/HtmlView.php [hunk 2]
```diff
@@ -79,13 +87,14 @@ public function display($tpl = null)
 		$this->filterForm    = $this->get('FilterForm');
 		$this->activeFilters = $this->get('ActiveFilters');
 
-		if (!count($this->items) && $this->get('IsEmptyState'))
+
+		if (!\count($this->items) && $this->isEmptyState = $this->get('IsEmptyState'))
```
Comment (richard67) [inline administrator/components/com_tags/src/View/Tags/HtmlView.php]:
```suggestion
		if (!\count($this->items) && $this->isEmptyState = $this->get('IsEmptyState'))
```

Comment (PhilETaylor) [reply administrator/components/com_tags/src/View/Tags/HtmlView.php]: I can't see any difference in that haha

Comment (richard67) [reply administrator/components/com_tags/src/View/Tags/HtmlView.php]: there were 2 empty lines in a row, the PHPCS doesn't like that

**Comment by richard67** [comment]
Previous tests and RTC are still valid since the 2 commits after the tests were only a clean update of the branch to the upstream 4.0-dev branch plus a code style fix.

**Comment by drmenzelit** [comment]
Thanks

**Pull request merged**